### PR TITLE
Add servlet for fetching contextual data on a new search term

### LINKED ISSUE #114: Add servlet for retrieving contextual data on a potential new investment

### PR DESCRIPTION
**Pull request by roonil-wazlib**
- minor bug fixes in the invesment calculator
 - add new servlet that either retrieves existing data from datastore and returns as array, or runs Cloud Function to fetch this data first.

Resolves #114

### CODE REVIEW

**Review by Phoebe-Khokgawe — APPROVED** [review]

File: project/src/main/java/com/google/sps/servlets/ContextDataServlet.java [hunk 1]
```diff
@@ -0,0 +1,101 @@
...
+        if (data == null) {
+            String date = calc.oneWeekBefore(calc.getLatestDate()).toString();
+
+            HashMap<String, String> arguments = new HashMap<>();
+            arguments.put("search", googleSearch);
+            arguments.put("date", date);
+
+            ByteString byteStr = ByteString.copyFrom(gson.toJson(arguments), StandardCharsets.UTF_8);
+
+            PubsubMessage pubsubApiMessage = PubsubMessage.newBuilder().setData(byteStr).build();
```
Comment (leonnemets) [inline project/src/main/java/com/google/sps/servlets/ContextDataServlet.java:64]: {nit} formatting

File: project/src/main/java/com/google/sps/servlets/ContextDataServlet.java [hunk 2]
```diff
@@ -0,0 +1,101 @@
...
+
+    @Override
+    public void doGet(HttpServletRequest request, HttpServletResponse response) throws IOException {
+        Gson gson = new Gson();
+        InvestmentCalculator calc = new InvestmentCalculator();
+
+        String googleSearch = request.getParameter("search_term");
+
+        ImmutableList<Long> data = calc.getInvestmentDataIfExists(googleSearch);
+        if (data == null) {
```
Comment (leonnemets) [inline project/src/main/java/com/google/sps/servlets/ContextDataServlet.java:55]: {Pet peeve} null arrays. Would be clearer to check for empty arrays

**Review by leonnemets — COMMENTED** [review]
There was discussion of how to implement this, CloudRun, pubsub etc. Do you have that documented anywhere coz the sleep while loop is unfortunate

File: project/src/main/java/com/google/sps/servlets/ContextDataServlet.java (hunk 2 above)
Comment (leonnemets) [reply project/src/main/java/com/google/sps/servlets/ContextDataServlet.java:55]: go/ej3e-54

Comment (roonil-wazlib) [reply project/src/main/java/com/google/sps/servlets/ContextDataServlet.java]: I have changed it

File: project/src/main/java/com/google/sps/servlets/InvestmentCalculator.java [hunk 3]
```diff
@@ -179,7 +180,8 @@ public int getInvestmentValue(String googleSearch, long investDate, long sellDat
             }
             return ImmutableList.copyOf(values);
         }
-        return null;
+        // no data was found
+        return ImmutableList.copyOf(values); //return empty array
```
Comment (leonnemets) [inline project/src/main/java/com/google/sps/servlets/InvestmentCalculator.java:184]: ImmutableList.of();

File: project/src/main/java/com/google/sps/servlets/ContextDataServlet.java [hunk 4]
```diff
@@ -0,0 +1,100 @@
...
+            // data retrieved from datastore, return data in json form
+            response.setContentType("application/json");
+            response.getWriter().println(gson.toJson(data));
+        }
+    }
+
+    /** 
+     * Check for data in datastore every 0.1 seconds until data exists or 15 seconds have elapsed
+     * Run this instead of waiting on cloud function return due to slight delay in datastore
+     */
```
Comment (leonnemets) [inline project/src/main/java/com/google/sps/servlets/ContextDataServlet.java:91]: from our discussion is this your theory? 
... slight delay in datastore being written and becoming readable?

**Review by leonnemets — APPROVED** [review]

File: project/src/main/java/com/google/sps/servlets/ContextDataServlet.java (hunk 4 above)
Comment (roonil-wazlib) [reply project/src/main/java/com/google/sps/servlets/ContextDataServlet.java:91]: I'm fairly certain that's the bug that I was getting before by waiting on a return.

File: project/src/main/java/com/google/sps/servlets/InvestmentCalculator.java (hunk 3 above)
Comment (roonil-wazlib) [reply project/src/main/java/com/google/sps/servlets/InvestmentCalculator.java]: sure

**Pull request merged**